### PR TITLE
amdvlk: 2020.Q3.4 -> 2020.Q3.5 and add 32-bit build

### LINKED ISSUE #93027: amdvlk 32bit
**Project description**
#82305 introduced amdvlk to nixos already, but afaict this only works with 64bit application, but since a lot of especially games are still 32bit it really would be nice to also be able to run those with amdvlk.

I currently have this:
```nix
{
environment.variables.VK_ICD_FILENAMES = "${pkgs.amdvlk}/share/vulkan/icd.d/amd_icd64.json:/run/opengl-driver-32/share/vulkan/icd.d/radeon_icd.i686.json";
}
```

In my nixos configuration to make sure that if possible amdvlk is used but also to make sure 32bit applications can still run

**Metadata**

should be the same

### PR DESCRIPTION
**Pull request by Flakebi**
###### Motivation for this change
AMDVLK 2020.Q3.5 was released: https://github.com/GPUOpen-Drivers/AMDVLK/releases/tag/v-2020.Q3.5
I took the opportunity to include 32-bit support (fixes #93027).

###### Things done

- [x] Tested with vkcube and Dota 2
- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/doc/manual/configuration/gpu-accel.xml [hunk 1]
```diff
@@ -184,6 +184,8 @@ GPU1:
 
 	<programlisting><xref linkend="opt-hardware.opengl.extraPackages"/> = [
   <package>amdvlk</package>
+  # For the 32-bit driver
+  <package>driversi686Linux.amdvlk</package>
```
Comment (danieldk) [inline nixos/doc/manual/configuration/gpu-accel.xml:188]: This should go in `hardware.opengl.extraPackages32`, right? Or doesn't the ICD loader pick it up there?

File: pkgs/development/libraries/amdvlk/default.nix [hunk 2]
```diff
@@ -62,14 +65,23 @@ stdenv.mkDerivation rec {
 
   cmakeDir = "../drivers/xgl";
 
+  # LTO is disabled in gcc for i686 as of #66528
+  cmakeFlags = if stdenv.system == "x86_64-linux" then [] else ["-DXGL_ENABLE_LTO=OFF"];
```
Comment (danieldk) [inline pkgs/development/libraries/amdvlk/default.nix:69]:
```suggestion
  cmakeFlags = stdenv.lib.optionals stdenv.is32bit ["-DXGL_ENABLE_LTO=OFF"];
```

File: pkgs/development/libraries/amdvlk/default.nix [hunk 3]
```diff
@@ -62,14 +65,23 @@ stdenv.mkDerivation rec {
...
+
+  postPatch = if stdenv.system == "x86_64-linux" then "" else ''
+    sed -i 's#pal_setup_gcc_ipo()##' drivers/pal/cmake/PalCompilerOptions.cmake
+  '';
```
Comment (danieldk) [inline pkgs/development/libraries/amdvlk/default.nix:73]:
```suggestion
  postPatch = stdenv.lib.optionalString stdenv.is32bit ''
    substituteInPlace drivers/pal/cmake/PalCompilerOptions.cmake \
      --replace "pal_setup_gcc_ipo()" ""
  '';
```

`substituteInPlace` is preferred, unless regexps are really needed. Also warns when no substitutions are made.

**Review by danieldk — CHANGES_REQUESTED** [review]

**Comment by danieldk** [comment]
Nice to have a 32-bit build as well!

File: nixos/doc/manual/configuration/gpu-accel.xml (hunk 1 above)
Comment (Flakebi) [reply nixos/doc/manual/configuration/gpu-accel.xml]: Ah, I didn’t know there was a 32-bit option too. Seems to work fine, 32-bit vkcube chose radv for me though 32-bit vulkaninfo showed both, radv and amdvlk.

**Comment by Flakebi** [comment]
Thanks for your comments, looks a lot cleaner!

File: nixos/doc/manual/configuration/gpu-accel.xml [hunk 4]
```diff
@@ -137,7 +137,7 @@ Platform Vendor      Advanced Micro Devices, Inc.</screen>
       <literal>:</literal>. For example:
 
       <screen><prompt>$</prompt> export \
-  VK_ICD_FILENAMES=`nix-build '&lt;nixpkgs&gt;' --no-out-link -A amdvlk`/share/vulkan/icd.d/amd_icd64.json</screen>
+  VK_ICD_FILENAMES=`nix-build '&lt;nixpkgs&gt;' --no-out-link -A amdvlk`/share/vulkan/icd.d/amd_icd64.json:`nix-build '&lt;nixpkgs&gt;' --no-out-link -A driversi686Linux.amdvlk`/share/vulkan/icd.d/amd_icd32.json</screen>
```
Comment (danieldk) [inline nixos/doc/manual/configuration/gpu-accel.xml:140]: I now looked at the rendered manual and this looks fairly complex as a first example as a first example of using `VK_ICD_FILENAMES`. I would just leave it as the original, most people should probably be using `extraPackages` anyway.

File: nixos/doc/manual/configuration/gpu-accel.xml [hunk 5]
```diff
@@ -182,7 +182,11 @@ GPU1:
 	available for applications and lets them choose. A specific driver can
 	be forced as follows:
 
-	<programlisting><xref linkend="opt-hardware.opengl.extraPackages"/> = [
+	<programlisting><xref linkend="opt-hardware.opengl.extraPackages"/> = with pkgs; [
+  <package>amdvlk</package>
+];
+
```
Comment (danieldk) [inline nixos/doc/manual/configuration/gpu-accel.xml:188]:
```suggestion

# To enable Vulkan support for 32-bit applications, also add:
```

File: nixos/doc/manual/configuration/gpu-accel.xml [hunk 6]
```diff
@@ -182,7 +182,11 @@ GPU1:
 	available for applications and lets them choose. A specific driver can
 	be forced as follows:
 
-	<programlisting><xref linkend="opt-hardware.opengl.extraPackages"/> = [
+	<programlisting><xref linkend="opt-hardware.opengl.extraPackages"/> = with pkgs; [
+  <package>amdvlk</package>
+];
```
Comment (danieldk) [inline nixos/doc/manual/configuration/gpu-accel.xml:187]:
```suggestion
	<programlisting><xref linkend="opt-hardware.opengl.extraPackages"/> = [
  pkgs.<package>amdvlk</package>
];
```

Comment (danieldk) [reply nixos/doc/manual/configuration/gpu-accel.xml:187]: Using `with` is discouraged. See:

https://github.com/NixOS/nixpkgs/pull/96960

File: nixos/doc/manual/configuration/gpu-accel.xml [hunk 7]
```diff
@@ -182,7 +182,11 @@ GPU1:
...
+
+<xref linkend="opt-hardware.opengl.extraPackages32"/> = with pkgs.driversi686Linux; [
```
Comment (danieldk) [inline nixos/doc/manual/configuration/gpu-accel.xml:189]: Here it would also be better to use

```
  pkgs.driversi686Linux<package>amdvlk</package>
```

**Review by danieldk — CHANGES_REQUESTED** [review]
Thanks for making the changes. Looks good now, only some small nitpicks about the manual changes.

**Review by danieldk — APPROVED** [review]
Thanks! LGTM

Result of `nixpkgs-review pr 97976` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>amdvlk</li>
    <li>driversi686Linux.amdvlk</li>
  </ul>
</details>

Will merge once ofborg is also happy.

**Pull request merged**